include/libcyphal/application/registry/register_impl.hpp

```diff
@@ -0,0 +1,350 @@
+/// @copyright
+/// Copyright (C) OpenCyphal Development Team  <opencyphal.org>
+/// Copyright Amazon.com Inc. or its affiliates.
+/// SPDX-License-Identifier: MIT
+
+#ifndef LIBCYPHAL_APPLICATION_REGISTRY_REGISTER_IMPL_HPP_INCLUDED
+#define LIBCYPHAL_APPLICATION_REGISTRY_REGISTER_IMPL_HPP_INCLUDED
+
+#include "registry_impl.hpp"
+#include "registry_value.hpp"
+
+#include <cetl/cetl.hpp>
+
+#include <utility>
+
+namespace libcyphal
+{
+namespace application
+{
+namespace registry
+{
+
+/// Defines abstract base class for a register implementation.
+///
+/// Implements common functionality for all register types like name, options, and value accessors.
+///
+class RegisterBase : public IRegister
+{
+    using Base = IRegister;
+
+public:
+    /// Defines options used when creating a new register.
+    ///
+    struct Options final
+    {
+        /// True if the register value is retained across application restarts.
+        bool persistent{false};
+
+    };  // Options
+
+    RegisterBase(const RegisterBase&)                      = delete;
+    RegisterBase& operator=(const RegisterBase&)           = delete;
+    RegisterBase& operator=(RegisterBase&& other) noexcept = delete;
+
+    /// Gets the register create options.
+    ///
+    Options getOptions() const noexcept
+    {
+        return options_;
+    }
+
+    // MARK: IRegister
+
+    Name getName() const override
+    {
+        return name_;
+    }
+
+protected:
+    RegisterBase(cetl::pmr::memory_resource& memory, const Name name, const Options& options)
+        : Base{name}
+        , allocator_{&memory}
+        , name_{name}
+        , options_{options}
+    {
+    }
+
+    ~RegisterBase()                       = default;
+    RegisterBase(RegisterBase&&) noexcept = default;
+
+    template <typename T>
+    ValueAndFlags getImpl(const T& value, const bool is_mutable) const
+    {
+        ValueAndFlags out{Value{allocator_}, {}};
+        registry::set(out.value_, value);
+        out.flags_.mutable_    = is_mutable;
+        out.flags_.persistent_ = options_.persistent;
+        return out;
+    }
+
+    template <typename T, typename Setter>
+    cetl::optional<SetError> setImpl(const T& value, Setter&& setter)
+    {
+        auto converted = get().value_;
+        if (coerce(converted, value))
+        {
+            if (std::forward<Setter>(setter)(converted))
+            {
+                return cetl::nullopt;
+            }
+            return SetError::Semantics;
+        }
+        return SetError::Coercion;
+    }
+
+private:
+    // MARK: Data members:
+
+    Value::allocator_type allocator_;
+    const Name            name_;
+    const Options         options_;
+
+};  // RegisterBase
+
+// MARK: -
+
+/// Defines a register implementation template.
+///
+template <typename Getter, typename Setter, bool IsMutable>
```

Comment:
Unless I'm misreading it, IsMutable seems to be redundant because you already have a specialization on `Setter=void` for that purpose